pallets/vesting/src/lib.rs

```diff
@@ -0,0 +1,271 @@
+
+	/// Information regarding the vesting of a given account.
+	#[pallet::storage]
+	#[pallet::getter(fn vesting_schedule)]
+	pub(super) type VestingSchedule<T: Config> = StorageValue<
+		_,
+		BoundedVec<(Percent, Schedule), T::MaxScheduleLength>,
+		ValueQuery,
+		DefaultVestingSchedule<T>,
+	>;
+
+	#[pallet::type_value]
+	pub(super) fn DefaultVestingSchedule<T: Config>(
+	) -> BoundedVec<(Percent, Schedule), T::MaxScheduleLength> {
+		BoundedVec::try_from(sp_std::vec![
+			(Percent::from_percent(34), 1636329600),
+			(Percent::from_percent(11), 1636502400),
+			(Percent::from_percent(11), 1641340800),
+			(Percent::from_percent(11), 1646179200),
+			(Percent::from_percent(11), 1651017600),
+			(Percent::from_percent(11), 1655856000),
+			(Percent::from_percent(11), 1660694400),
+		])
+		.unwrap_or_default()
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// The amount vested has been updated. This could indicate more funds are available. The
+		/// balance given is the amount which is left unvested (and thus locked).
+		/// \[account, unvested\]
+		VestingUpdated(T::AccountId, BalanceOf<T>),
+		/// An \[account\] has become fully vested. No further vesting can happen.
+		VestingCompleted(T::AccountId),
+		/// Update a vesting schedule.
+		VestingScheduleUpdated(BoundedVec<Schedule, T::MaxScheduleLength>),
+	}
+
+	/// Error for the vesting pallet.
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The account given is not vesting.
+		NotVesting,
+		/// An existing vesting schedule already exists for this account that cannot be clobbered.
+		ExistingVestingSchedule,
+		/// Amount being transferred is too low to create a vesting schedule.
+		AmountLow,
+		/// Not enough tokens for vesting.
+		BalanceLow,
+		/// Cannot input
+		InvalidTimestamp,
+		/// The length of new schedule cannot be bigger/smaller than 7.
+		InvalidScheduleLength,
+		/// The new schedule should be sorted.
+		UnsortedSchedule,
+		/// The first round of vesting is not done yet.
+		ClaimTooEarly,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Update vesting schedule.
+		///
+		/// - `new_schedule`: New schedule for vesting.
+		#[pallet::weight(10_000)]
+		pub fn update_vesting_schedule(
+			origin: OriginFor<T>,
+			new_schedule: BoundedVec<Schedule, T::MaxScheduleLength>,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+
+			let old_schedule = VestingSchedule::<T>::get();
+			ensure!(
+				new_schedule.len() == old_schedule.len(),
+				Error::<T>::InvalidScheduleLength
+			);
+
+			ensure!(
+				new_schedule.as_slice().windows(2).all(|w| w[0] < w[1]),
+				Error::<T>::UnsortedSchedule
+			);
+
+			let now = T::Timestamp::now().as_secs();
+			ensure!(
+				new_schedule.iter().all(|&s| now <= s),
+				Error::<T>::InvalidTimestamp
+			);
```

Comment:
The check is incorrect here, say if we want to update schedule after the first vesting happens. This check will fail.

We should check the following:
1. In a new schedule, if that is a past time, it should be consistent with the old_schedule
2. if that is a future time, it is fine. We also need to put some `buffer` of future time considering the block finality.

Reply:
Also, please update the logic of this check to the rustdoc comments.